core/arch/arm/kernel/stmm_sp.c

```diff
@@ -818,30 +841,109 @@ static void stmm_handle_storage_service(struct thread_scall_regs *regs)
 	service_compose_direct_resp(regs, stmm_rc);
 }
 
-static void spm_eret_error(int32_t error_code, struct thread_scall_regs *regs)
+static void spm_handle_direct_req(struct thread_scall_regs *regs)
 {
-	SVC_REGS_A0(regs) = FFA_ERROR;
+	uint16_t dst_id = SVC_REGS_A1(regs) & UINT16_MAX;
+
+	if (dst_id == ffa_storage_id) {
+		stmm_handle_storage_service(regs);
+	} else {
+		EMSG("Undefined endpoint id %#"PRIx16, dst_id);
+		spm_eret_error(STMM_RET_INVALID_PARAM, regs);
+	}
+}
+
+static void spm_handle_get_mem_attr(struct thread_scall_regs *regs)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct ts_session *sess = NULL;
+	struct stmm_ctx *spc = NULL;
+	uint16_t attrs = 0;
+	uint16_t perm = 0;
+	vaddr_t va = 0;
+	uint32_t ffa_ret = FFA_INVALID_PARAMETERS;
+
+	sess = ts_get_current_session();
+	spc = to_stmm_ctx(sess->ctx);
+
+	va = SVC_REGS_A1(regs);
+	if (!va)
+		goto err;
+
+	res = vm_get_prot(&spc->uctx, va, SMALL_PAGE_SIZE, &attrs);
+	if (res)
+		goto err;
+
+	if ((attrs & TEE_MATTR_URW) == TEE_MATTR_URW)
+		perm |= FFA_MEM_PERM_RW;
+	else if ((attrs & TEE_MATTR_UR) == TEE_MATTR_UR)
+		perm |= FFA_MEM_PERM_RO;
+
+	if (!(attrs & TEE_MATTR_UX))
+		perm |= FFA_MEM_PERM_NX;
+
+	SVC_REGS_A0(regs) = __FFA_SUCCESS;
 	SVC_REGS_A1(regs) = FFA_PARAM_MBZ;
-	SVC_REGS_A2(regs) = error_code;
+	SVC_REGS_A2(regs) = perm;
 	SVC_REGS_A3(regs) = FFA_PARAM_MBZ;
 	SVC_REGS_A4(regs) = FFA_PARAM_MBZ;
 	SVC_REGS_A5(regs) = FFA_PARAM_MBZ;
 	SVC_REGS_A6(regs) = FFA_PARAM_MBZ;
 	SVC_REGS_A7(regs) = FFA_PARAM_MBZ;
+
+	return;
+
+err:
+	spm_eret_error(ffa_ret, regs);
 }
 
-static void spm_handle_direct_req(struct thread_scall_regs *regs)
+static void spm_handle_set_mem_attr(struct thread_scall_regs *regs)
 {
-	uint16_t dst_id = SVC_REGS_A1(regs) & UINT16_MAX;
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct ts_session *sess = NULL;
+	struct stmm_ctx *spc = NULL;
+	uintptr_t va = SVC_REGS_A1(regs);
+	uint32_t nr_pages = SVC_REGS_A2(regs);
+	uint32_t perm = SVC_REGS_A3(regs);
+	size_t sz = 0;
+	uint32_t prot = 0;
+	uint32_t ffa_ret = FFA_INVALID_PARAMETERS;
 
-	if (dst_id == mem_mgr_id) {
-		stmm_handle_mem_mgr_service(regs);
-	} else if (dst_id == ffa_storage_id) {
-		stmm_handle_storage_service(regs);
-	} else {
-		EMSG("Undefined endpoint id %#"PRIx16, dst_id);
-		spm_eret_error(STMM_RET_INVALID_PARAM, regs);
+	if (!va || !nr_pages ||
+	    MUL_OVERFLOW(nr_pages, SMALL_PAGE_SIZE, &sz) ||
+	    (perm & FFA_MEM_PERM_RESERVED))
+		goto err;
+
+	sess = ts_get_current_session();
+	spc = to_stmm_ctx(sess->ctx);
+
+	if ((perm & FFA_MEM_PERM_DATA_PERM) == FFA_MEM_PERM_RO)
+		prot |= TEE_MATTR_UR;
+	else if ((perm & FFA_MEM_PERM_DATA_PERM) == FFA_MEM_PERM_RW)
+		prot |= TEE_MATTR_URW;
+
+	if ((perm & FFA_MEM_PERM_INSTRUCTION_PERM) != FFA_MEM_PERM_NX)
+		prot |= TEE_MATTR_UX;
+
+	res = vm_set_prot(&spc->uctx, va, sz, prot);
+	if (res) {
+		ffa_ret = FFA_DENIED;
+		goto err;
 	}
+
+	SVC_REGS_A0(regs) = __FFA_SUCCESS;
```

Comment:
`FFA_SUCCESS64` is used only if any result register encodes a 64-bit parameter. That's not the case here. Please fix the other places too.